apps/coscrad-frontend/src/components/resources/vocabulary-lists/vocabulary-list-detail.full-view.presenter.tsx

```diff
@@ -1,21 +1,101 @@
-import { IDetailQueryResult, IVocabularyListViewModel } from '@coscrad/api-interfaces';
+import {
+    IDetailQueryResult,
+    IVocabularyListEntry,
+    IVocabularyListViewModel,
+} from '@coscrad/api-interfaces';
+import { useReducer } from 'react';
 import { Carousel } from '../../higher-order-components/carousel';
 import { TermDetailFullViewPresenter } from '../terms/term-detail.full-view.presenter';
+import doValuesMatchFilters from './do-values-match-filters';
 import { formatBilingualText } from './utils';
+import { VocabularyListForm } from './vocabulary-list-form';
+
+type VocabularyListFilterProperty = string | boolean;
+
+export type VocabularyListFilter = Record<string, VocabularyListFilterProperty>;
+
+const UPDATE_VOCABULARY_LIST_FILTER = 'UPDATE_VOCABULARY_LIST_FILTER';
+
+type UpdateVocabularyListFilterPayload = {
+    key: string;
+    value: string | boolean;
+};
+
+type FSA<T, U> = {
+    type: T;
+    payload: U;
+};
+
+const updateVocabularyListFilter = (
+    key: string,
+    value: VocabularyListFilterProperty
+): FSA<typeof UPDATE_VOCABULARY_LIST_FILTER, UpdateVocabularyListFilterPayload> => ({
+    type: UPDATE_VOCABULARY_LIST_FILTER,
+    payload: {
+        key,
+        value,
+    },
+});
+
+const filterReducer = (
+    state: VocabularyListFilter,
+    {
+        type,
+        payload: { key, value },
```

Comment:
This is like the command system on the api except used with redux?

Reply:
More technically, the reducer pattern is the same pattern used by Redux. `useReducer` allows you to use the reducer pattern to manage state (`useState` + reducer pattern).

`useReducer` and Redux both use the `Flux Standard Action` standard for communicating "Actions". It is also conventional to use this standard in CQRS systems, on which I based our `@coscrad/commands` lib.

So, really, Redux is just event sourcing front-end in-memory state!!! You could imagine sending the domain models to the frontend, converting our commands to actions, and processing commands in-memory! Then you'd have to have  a way to merge event streams \ reconcile conflicts.